readme.rst
Jirafs
======

.. image:: https://travis-ci.org/coddingtonbear/jirafs.svg?branch=master
    :target: https://travis-ci.org/coddingtonbear/jirafs

.. image:: https://badge.fury.io/py/jirafs.png
    :target: http://badge.fury.io/py/jirafs

Pronounced like 'giraffes', but totally unrelated to wildlife, this
library lets you stay out of Jira as much as possible by letting
you edit your Jira issues as a collection of text files using an
interface inspired by ``git`` and ``hg``.

.. image:: http://coddingtonbear-public.s3.amazonaws.com/github/jirafs/readme_demo_2.1_v7.gif

Read more details in `the docs <http://jirafs.readthedocs.org/>`_.

Plugins and Macros Too
----------------------

Jirafs provides a Plugin API allowing you to simplify your workflow in Jira;
several plugins already exist, including the following tools:

* For including programmatically-generated images in your Jira issues without
  ever leaving your editor:

  * `jirafs-graphviz <http://github.com/coddingtonbear/jirafs-graphviz>`_:
    Embed Graphviz (e.g. ``dot`` or ``neato``) graphs using Graphviz's
    ubiquitous graph description language.
  * `jirafs-matplotlib <http://github.com/coddingtonbear/jirafs-matplotlib>`_:
    Embed graphs generated with the common Python charting library Matplotlib
    by writing simple python scripts.
  * `jirafs-plantuml <http://github.com/coddingtonbear/jirafs-plantuml>`_:
    Embed UML (e.g. timing, sequence, or activity) diagrams
    generated via PlantUML's easy-to-use text format.
  * `jirafs-mermaid <http://github.com/coddingtonbear/jirafs-mermaid>`_:
    Embed beautiful diagrams (e.g. pie, gantt, or class)
    using Mermaid's markdown-ish diagram description language.

* For making tables more easily:

  * `jirafs-csv-table <http://github.com/coddingtonbear/jirafs-csv-table>`_:
    Include tables in Jira by generating them from local CSV files.
  * `jirafs-list-table <http://github.com/coddingtonbear/jirafs-list-table>`_:
    Create tables in Jira by using a simple list-based syntax.

Requirements
------------

* ``git >= 1.8``
* ``python3 >= 3.6``

----------

- Documentation for Jirafs is available on
  `ReadTheDocs <http://jirafs.readthedocs.org/>`_.
- Please post issues on
  `Github <http://github.com/coddingtonbear/jirafs/issues>`_.
- Test status available on
  `Travis-CI <https://travis-ci.org/coddingtonbear/jirafs>`_.
- Questions? Ask them on
  `Gitter <https://gitter.im/coddingtonbear/jirafs>`_.


CHANGELOG.rst
2.1.4
-----

* Fixes a bug in which search results from ``search_users`` were not
  being displayed.  (Issue #61)
* Fixes another field deletion/addition problem relating to Issue #59
  (see release notes for 2.1.3 for more).

2.1.3
-----

* Fixes a bug preventing the ``match`` subcommand from being able
  to properly obtain field values.  (Issue #60)
* Fixes a bug that would cause an exception to be raised when calculating
  field differences if one's Jira instance either added or removed fields
  between different ticketfolder versions.  (Issue #59)

2.1.2
-----

* Fixes several minor bugs relating to issue previews including:

  * In certain situations, the preview automatic change detection
    would become stuck in a loop causing the page to refresh
    continuously.
  * Preview document was not delivered with an UTF-8 charset,
    so non-latin-1 characters would be mangled.
  * Changes to Jira fields would not be refreshed when rendering
    preview after submitting changes.
  * Insufficient whitespace between comments may cause Jira to
    misinterpret headers as part of a previous comment's bulleted
    list.

2.1.1
-----

* Fixes bug that would cause an unhelpful traceback to be displayed if you
  were to run Jirafs commands outside of a ticketfolder.

2.1.0
-----

* Fixes to various bugs relating to merging upstream changes
  from Jira with un-pushed local changes.

  * User is now warned about conflicts were they to occur when merging-in
    upstream changes.
  * Conflicts are displayed on the ``status`` display should they exist,
    including instructions about how to resolve them.

* Additional polish for 'preview' function including:

  * Fixes to exception display in-console for harmless errors resulting
    from clients navigating away.
  * Defaulting behavior such that server will automatically self-terminate
    when the user navigates away.  If this behavior is not desired,
    a command-line argument ``--serve-forever`` is available.
  * Adding display of comments for 'all' preview mode.
  * Adds links allowing you to jump to the specific section you're interested
    in when in the 'all' preview mode.

* Macro reversal changes for ``comments.read_only.jira``: No longer will
  macros be reversed for read-only files given that the content of
  historical comments cannot be changed.
* Users can now set a preferred date format by setting a configuration
  value ``main.date_format``.
* Obsolete and undocumented "Clone" functionality allowing you to clone
  a Jira ticketfolder from a git repository (instead of directly from
  a Jira issue) was removed.
* Jira server URL inference when cloning by using an issue number instead
  of a ticket URL is improved to be more foolproof.

2.0.0
-----

* Major revamping of Plugin API to allow for more-reliable
  macro application and reversal, among other optimizations.
* New 'preview' command allowing you to preview content entered
  in any field for quickly verifying formatting and macro-generated content.
* Changes & improvements across the board.

1.2.0
-----

* Adds new ``match`` command allowing you to test whether fields match (or
  do not match) expected values.  This is particularly useful for running
  commands on any ticket folders in a specific folder that might match a
  certain condition (e.g: that they're closed).
* Cleanup of existing internal plugins such that the ``main`` method
  performs the majority of the task's work instead of the method named
  after the command's name.
* Added functionality to existing plugin architecture allowing one to
  control the return value of the command.

1.13.2
------

* Fixing a bug that prevented multiple macros from successfully being
  applied to a file.

1.13.1
------

* Patch version bump because I ran the release for 1.13.0 from the wrong
  branch, so it was missing some minor bugfixes.

1.13.0
------

* Changes functionality of "Macro Plugins" to apply only to file-based
  fields.  This will prevent misbehaving macros from making the general
  field file from becoming conflicted.

1.12.0
------

* Adds enable/disable functionality to macro plugins.
* Moved ``{list-table}`` plugin out into its own repository.  To install this
  macro again, see https://github.com/coddingtonbear/jirafs-list-table.
* Adds new ``macropatch`` command allowing you to reset or see the existing
  macropatch.  This is especially important for use in situations where you
  disable a macro plugin.
* Fixed a bug in config handling where on Python 2.7 configparser would
  raise an error when attempting to enable a plugin.

1.11.0
------

* Adds new "Macro Plugins" support including a new built-in macro --
  ``{list-table}`` used for making tables a little bit simpler.
* Alters ``field`` command to return text *after* being processed
  by macros.  Use the ``--raw`` argument to instead return the
  pre-processed field contents.

1.10.0
------

* Adds new ``create`` command providing you with a way of interactively
  creating a new issue.  Note that ``create`` will automatically
  clone the issue once you've supplied the requisite informaiton.
* Adds new ``submit`` command which runs a ``commit`` followed by
  a ``push`` (which itself automatically runs ``pull``); although
  the lower-level git-like commands are still available, using ``submit``
  is now the preferred method of interacting with Jirafs.

1.9.0
-----

* Adds new ``search_users`` command allowing one to search for users.  This
  is particularly useful for when attempting to mention somebody in a ticket
  comment, but you're not sure what their user name is.
* When running ``merge`` (even via ``pull``) messages will be displayed
  indicating remote changes that are being merged-in to your working copy.
* Adds basic integration tests; this should add a lot of insulation preventing
  me (or anybody else) from accidentally breaking Jirafs for versions of
  Python not in use by the writer.

1.8.0
-----


jirafs/cmdline.py
import argparse
import codecs
import copy
import logging
import logging.config
import os
import shlex
import subprocess
import sys
import time
import traceback

from blessings import Terminal

try:
    from jira.utils import JIRAError
except ImportError:
    from jira.exceptions import JIRAError
from distutils.version import LooseVersion

from . import utils
from .exceptions import (
    GitCommandError,
    JiraInteractionFailed,
    JirafsError,
    NotTicketFolderException,
    UnknownMacroError,
    MacroError,
)


# Write data to stdout as UTF-8 bytes when there's no encoding specified
if sys.version_info < (3,) and sys.stdout.encoding is None:
    sys.stdout = codecs.getwriter("utf-8")(sys.stdout)


logger = logging.getLogger(__name__)


LOGGING = {
    "version": 1,
    "disable_existing_loggers": False,
    "formatters": {
        "pretty": {"format": "[%(levelname)s] %(message)s"},
        "file": {"format": "%(asctime)s\t%(levelname)s\t%(module)s\t%(message)s"},
    },
    "handlers": {
        "console": {
            "level": "INFO",
            "class": "logging.StreamHandler",
            "formatter": "pretty",
            "stream": "ext://sys.stdout",
        }
    },
    "root": {"handlers": ["console"], "level": "INFO"},
}


def main():
    term = Terminal()
    if sys.version_info < (2, 7):
        raise RuntimeError(
            "Jirafs requires minimally version 2.7 of Python 2, or "
            "any version of Python 3.  Please upgrade your version of "
            "python before using Jirafs."
        )
    if utils.get_git_version() < LooseVersion("1.8"):
        raise RuntimeError(
            "Jirafs requires minimally version 1.8 of Git.  Please "
            "upgrade your version of git before using Jirafs."
        )

    commands = utils.get_installed_commands()

    parser = argparse.ArgumentParser(
        description="Edit Jira issues locally from your filesystem", add_help=False,
    )
    parser.add_argument("command", type=str, choices=commands.keys())
    parser.add_argument("--subtasks", action="store_true", default=False)
    parser.add_argument(
        "--log-level", default="INFO", dest="log_level",
    )
    parser.add_argument("--folder", default=os.getcwd())
    parser.add_argument(
        "--no-subfolders", action="store_true", default=False,
    )
    parser.add_argument(
        "--traceback", action="store_true", default=False,
    )
    parser.add_argument(
        "--profile",
        action="store_true",
        default=False,
        help="Generate a stack profile for this operation; requires 'pyinstrument'.",
    )
    parser.add_argument(
        "--debugger-port",
        default=58024,
        help=("(Requires --debugger) start debugger on this port."),
    )
    parser.add_argument(
        "--debugger",
        action="store_true",
        default=False,
        help=("Launch ptvsd debugger on --debugger-port."),
    )
    args, extra = parser.parse_known_args()

    if args.debugger:
        try:
            import ptvsd

            ptvsd.enable_attach(address=("localhost", args.debugger_port))
            print(
                f"{term.magenta}Waiting for debugger connection on "
                f"port {args.debugger_port}...{term.normal}"
            )
            ptvsd.wait_for_attach()
        except ImportError:
            print(f"{term.red}Module 'ptvsd' required for debugging.{term.normal}")
            sys.exit(1)

    if args.profile:
        try:
            import pyinstrument

            profiler = pyinstrument.Profiler()
            profiler.start()
        except ImportError:
            print(
                f"{term.red}Module 'pyinstrument' required for profiling."
                f"{term.normal}"
            )
            sys.exit(1)

    logging.config.dictConfig(LOGGING)
    root_logger = logging.getLogger()
    root_logger.setLevel(logging.getLevelName(args.log_level))

    command_name = args.command
    cmd_class = commands[command_name]

    # Subtasks
    if args.subtasks:
        cmd_class.RUN_FOR_SUBTASKS = True

    started = time.time()
    logger.debug("Command %s(%s) started", command_name, extra)
    jira = utils.lazy_get_jira()
    try:
        value = cmd_class.execute_command(
            extra, jira=jira, path=args.folder, command_name=command_name
        )
        logger.debug(
            "Command %s(%s) finished in %s seconds",
            command_name,
            extra,
            (time.time() - started),
        )
        if value:
            value.echo()
        sys.exit(value.return_code)
    except GitCommandError as e:
        print(
            "{t.red}Error (code: {code}) while running git "
            "command.{t.normal}".format(t=term, code=e.returncode)
        )
        print("")
        print("{t.red}Command:{t.normal}{t.red}{t.bold}".format(t=term))
        print("    {cmd}".format(cmd=e.command))
        print("{t.normal}".format(t=term))
        print("{t.red}Output:{t.normal}{t.red}{t.bold}".format(t=term))
        for line in e.output.decode("utf8").split("\n"):
            print("    %s" % line)
        print("{t.normal}".format(t=term))
        if args.traceback:
            traceback.print_exc()
        sys.exit(10)
    except NotTicketFolderException:
        if not getattr(cmd_class, "TRY_SUBFOLDERS", False):
            print(
                "{t.red}The command '{cmd}' must be ran from "
                "within an issue folder.{t.normal}".format(t=term, cmd=command_name)
            )
            sys.exit(20)
        elif args.no_subfolders:
            sys.exit(20)

        count_runs = 0
        for folder in os.listdir(os.getcwd()):
            full_path = os.path.join(os.getcwd(), folder,)
            if not os.path.isdir(full_path):
                continue

            try:
                full_args = copy.copy(sys.argv)
                if "--no-subfolders" not in full_args:
                    full_args.append("--no-subfolders")
                result = subprocess.call(
                    " ".join([shlex.quote(a) for a in full_args]),
                    cwd=full_path,
                    shell=True,
                )
                if result == 0:
                    count_runs += 1
            except NotTicketFolderException:
                pass
        if count_runs == 0:
            if args.traceback:
                traceback.print_exc()
            sys.exit(21)
    except UnknownMacroError as e:
        print(
            "{t.red}Jirafs encountered an unknown macro while processing "
            "your content: {t.normal}{t.red}{t.bold}{error}"
            "{t.normal}".format(t=term, error=str(e))
        )
        if args.traceback:
            traceback.print_exc()
        sys.exit(30)
    except MacroError as e:
        print(
            "{t.red}Jirafs encountered an error while processing a "
            "{t.bold}{macro}{t.normal}{t.red} macro: "
            "{t.normal}{t.red}{t.bold}{error}{t.normal}".format(
                t=term, macro=e.macro_name or "?", error=str(e)
            )
        )
        if args.traceback:
            traceback.print_exc()
        sys.exit(40)
    except JIRAError as e:
        print(
            "{t.red}Jirafs encountered an error while interacting with "
            "your Jira instance: {t.normal}{t.red}{t.bold}{error}"
            "{t.normal}".format(t=term, error=str(e))
        )
        if args.traceback:
            traceback.print_exc()
        sys.exit(70)
    except JiraInteractionFailed as e:
        print(
            "{t.red}Jira was unable to satisfy your "
            "request: {t.normal}{t.red}{t.bold}{error}{t.normal}".format(
                t=term, error=str(e)
            )
        )
        if args.traceback:
            traceback.print_exc()
        sys.exit(80)
    except JirafsError as e:
        print(
            "{t.red}Jirafs encountered an error processing your "
            "request: {t.normal}{t.red}{t.bold}{error}{t.normal}".format(
                t=term, error=str(e)
            )
        )
        if args.traceback:
            traceback.print_exc()
        sys.exit(90)
    finally:
        if args.profile:
            profiler.stop()
            print(profiler.output_text(unicode=True, color=True))
